velox/exec/Window.cpp

```diff
@@ -282,6 +314,131 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
   }
 }
 
+void Window::updateKRowsFrameBounds(
+    const bool isKPreceding,
+    const FrameChannelArg& frameArg,
+    const vector_size_t startRow,
+    const vector_size_t numRows,
+    vector_size_t* rawFrameBounds) {
+  auto firstPartitionRow = partitionStartRows_[currentPartition_];
+
+  if (frameArg.index == kConstantChannel) {
+    auto constantOffset = frameArg.constant.value();
+    std::iota(
+        rawFrameBounds,
+        rawFrameBounds + numRows,
+        startRow + (isKPreceding ? -constantOffset : constantOffset) -
+            firstPartitionRow);
+  } else {
+    frameArg.value->resize(numRows);
+    windowPartition_->extractColumn(
+        frameArg.index, partitionOffset_, numRows, 0, frameArg.value);
+    auto offsets = frameArg.value->values()->as<int64_t>();
+    for (auto i = 0; i < numRows; i++) {
+      if (offsets[i] <= 0) {
+        VELOX_USER_FAIL(
+            "{} in frame bound at index {} must be at least 1", offsets[i], i);
```

Comment:
The first {} is the row number which comes from any point in the partition. So let's remove it. The error message could be "Frame bound at index {} contains a row with value < 1"

Reply:
Thanks for the suggestion, the first parameter gives the value at the row and was added for ease of debugging. Error message is updated accordingly.